hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,523 @@
+import warnings
+from collections import Counter, defaultdict
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from .._ipcw import KaplanMeierIPCW
+from ..utils import check_y_survival
+
+
+def concordance_index_incidence(
+    y_test,
+    y_pred,
+    y_train=None,
+    ipcw_estimator="km",
+    time_grid=None,
+    taus=None,
+    event_of_interest=1,
+    tied_tol=1e-8,
+):
+    r"""Time-dependent concordance index for prognostic models with competing risks \
+    using inverse probability of censoring weighting.
+
+    This C-index is event-specific: we set the event of interest as a parameter,
+    and all other event labels are considered to be competing events –except 0, which
+    is the censoring event.
+
+    For example, if ``y_test`` has the following ``"event"`` columns: ``[1, 0, 2, 4]``,
+    and the event of interest is set to ``2``, then we consider the following events:
+    ``[2, 0, 1, 2]``, where 0, 1, 2 respectively denotes the censoring event, the event
+    of interest and competing events.
+
+    We define this C-index as:
+
+    .. math::
+
+        \mathrm{C}(t) = \frac{\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) Q^{ij}(t)
+        \tilde{N}^1_i(t)}
+        {\sum_{i=1}^n \sum_{j=1}^n (\tilde{A}_{ij}
+        \hat{W}_{ij, 1}^{-1} + \tilde{B}_{ij} \hat{W}_{ij, 2}^{-1}) \tilde{N}^1_i(t)}
+
+    where:
+
+    .. math::
+
+        \begin{align}
+        \tilde{N}^1_i(t) &= I\{\tilde{T}_i \leq t, \tilde{D}_i = 1\} \\
+        \tilde{A}_{ij} &= I\{\tilde{T}_i < \tilde{T}_j \cup (\tilde{T}_i =
+        \tilde{T}_j \cap D_j = 0)\} \\
+        \tilde{B}_{ij} &= I\{\tilde{T}_i \geq \tilde{T}_j, D_j = 2\} \\
+        \hat{W}_{ij,1} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_i|X_j) \\
+        \hat{W}_{ij,2} &= \hat{G}(\tilde{T}_i-|X_i) \hat{G}(\tilde{T}_j-|X_j) \\
+        Q_{ij}(t) &= I\{M(t, X_i) > M(t, X_j)\}
+        \end{align}
+
+    where:
+
+    - :math:`\tilde{T_i} = \min(T_i, C_i)` and
+      :math:`\tilde{D_i} = I\{T_i \leq C_i \} D_i` are the observed time-to-event and
+      the observed event.
+    - :math:`D_j = 0`, :math:`D_j = 1` and :math:`D_j = 2` respectively indicate
+      individuals having been censored, individuals having experienced the event of
+      interest, and individual having experienced a competing event.
```

Comment:
You could shorten a little bit
```suggestion
    - :math:`D_j = 0, 1, 2` respectively denotes a censoring event, 
    the event of interest and competing events.
```